pkg/distributor/ha_tracker.go

```diff
@@ -472,22 +487,37 @@ func (h *haTracker) updateCache(userID, cluster string, desc *ReplicaDesc) {
 func (h *haTracker) updateKVStore(ctx context.Context, userID, cluster, replica string, now time.Time) error {
 	key := fmt.Sprintf("%s/%s", userID, cluster)
 	var desc *ReplicaDesc
+	var electedAtTime, electedChanges int64
 	err := h.client.CAS(ctx, key, func(in interface{}) (out interface{}, retry bool, err error) {
 		var ok bool
 		if desc, ok = in.(*ReplicaDesc); ok && desc.DeletedAt == 0 {
 			// If the entry in KVStore is up-to-date, just stop the loop.
-			if h.withinUpdateTimeout(now, desc.ReceivedAt) ||
-				// If our replica is different, wait until the failover time.
-				desc.Replica != replica && now.Sub(timestamp.Time(desc.ReceivedAt)) < h.cfg.FailoverTimeout {
+			if h.withinUpdateTimeout(now, desc.ReceivedAt) {
 				return nil, false, nil
 			}
+			electedAtTime = desc.ElectedAt
+			electedChanges = desc.ElectedChanges
+			// If our replica is different, wait until the failover time
+			if desc.Replica != replica {
+				if now.Sub(timestamp.Time(desc.ReceivedAt)) < h.cfg.FailoverTimeout {
+					level.Info(h.logger).Log("msg", "replica differs, but it's too early to failover", "user", userID, "cluster", cluster, "replica", replica, "elected", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
+					return nil, false, nil
+				}
+				level.Info(h.logger).Log("msg", "replica differs, attempting to update kv", "user", userID, "cluster", cluster, "replica", replica, "elected", desc.Replica, "received_at", timestamp.Time(desc.ReceivedAt))
+				electedAtTime = timestamp.FromTime(now)
+				electedChanges = desc.ElectedChanges + 1
+			}
+		}
+		if electedAtTime == 0 {
+			electedAtTime = timestamp.FromTime(now)
 		}
```

Comment:
Note that if `desc.DeletedAt != 0` the we'd be updating the `electedAtTime` here and re-setting the `electedChanges`. I think we should keep those.